themes/default/content/learn/abstraction-encapsulation/encapsulation/index.md

```diff
@@ -0,0 +1,150 @@
+import pulumi
+import pulumi_aws_native as aws
+
+bucket = aws.s3.bucket("my-bucket")
+
+pulumi.export("bucket", bucket.bucket_name)
+```
+
+Here, we're creating one resource and exporting the output from that resource.
+Generally, however, there's a lot more to building up infrastructure with a
+storage object. We need access policies, for one. Depending on how you're using
+that storage, you might need networking and other resources, as well. All of
+these use cases for a storage object requires some kind of policy, though, so
+when thinking about encapsulating our storage object needs, we probably want to
+put those two together. Let's adjust our code accordingly:
+
+```python
+import json
+import pulumi
+import pulumi_aws as aws_classic
+import pulumi_aws_native as aws
+
+bucket = aws.s3.bucket("my-bucket")
+bucket_policy = aws_classic.s3.BucketPolicy(
+    "my-bucket-policy",
+    bucket=bucket.id,
+    policy=bucket.arn.apply(
+        lambda arn: json.dumps({
+            "Version": "2012-10-17",
+            "Statement": [{
+                "Effect": "Allow",
+                "Principal": "*",
+                "Action": [
+                    "s3:GetObject"
+                ],
+                "Resource": [
+                    f"{arn}/*"
+                ]
+            }]
+        })
+    ),
+    opts=pulumi.ResourceOptions(parent=bucket)
+)
+
+pulumi.export("bucket", bucket.bucket_name)
+```
+
+Now, that's nice, but what if we need to make three different storage objects
+for three different use cases? We could start out by creating a resource
+grouping for the kind of resource we wanted to use:
+
+```python
+import ...
+
+
+class OurBucketClass(self, name_me):
+    bucket = aws.s3.bucket(f"{name_me}")
+    bucket_policy = aws_classic.s3.BucketPolicy(
+        f"{name_me}-policy",
+        bucket=bucket.id,
+        policy=bucket.arn.apply(
+            lambda arn: json.dumps({
+                "Version": "2012-10-17",
+                "Statement": [{
+                    "Effect": "Allow",
+                    "Principal": "*",
+                    "Action": [
+                        "s3:GetObject"
+                    ],
+                    "Resource": [
+                        f"{arn}/*"
+                    ]
+                }]
+            })
+        ),
+        opts=pulumi.ResourceOptions(parent=bucket)
+    )
+```
+
+But that JSON list of the policy is an object, too, that we can tuck away,
+making our storage object class more usable outside of this specific context.
+Let's imagine we have a referenceable list of common access policies, perhaps
+a simple key:value store. In the store, keys are strings that are names of
+policies, and the values are JSON documents.
+
+```python
+import ...
+
+
+class OurBucketClass(self, name_me, policy_name):
+
+    policy_list = {
+        'default': '{...}',
+        'locked' : '{...}',
+        'permissive': '{...}'
+    }
+
+    def define_policy(policy_name, bucket_id):
+        try:
```

Comment:
This `try` should be removed as it has no effect since the `except` block has no error handling.

Reply:
whoops, forgot that; I'm fixing the block with better handling in the next commit